### PR TITLE
cmd/libsnap-confine-private: introduce a helper for splitting snap name

### PR DESCRIPTION
**Pull request by bboozzoo**
Introduce a helper for splitting snap name into the name and instance key parts.

### CODE REVIEW

File: cmd/libsnap-confine-private/snap.h [hunk 1]
```diff
@@ -71,4 +71,19 @@ bool sc_is_hook_security_tag(const char *security_tag);
...
+/**
+ * Extract snap name and instance key out of a snap with optional instance name string.
+ *
+ * A snap may be installed multiple times in parallel under distinct instance
+ * names. This function extracts the snap name and instance key out of a name
+ * that possibly contains a snap instance name.
+ *
+ * For example:
+ *    snap_instance => "snap"      & "instance"
+ *    just-snap     => "just-snap" & ""
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.h]: Can you update the documentation to include the fact that both base and instance can be NULL.

File: cmd/libsnap-confine-private/snap.h [hunk 2]
```diff
@@ -71,4 +71,19 @@ bool sc_is_hook_security_tag(const char *security_tag);
...
+ **/
+void sc_snap_split_instance_name(const char *snap_name, char *base,
+				 size_t base_size, char *instance_key,
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.h]: Can we find a better name not to use the word _base_ here? How about "full name" for snap+instance and just snap_name and instance_key?

File: cmd/libsnap-confine-private/snap.c [hunk 3]
```diff
@@ -172,22 +172,43 @@ void sc_snap_name_validate(const char *snap_name, struct sc_error **errorp)
...
+			die("base buffer too small");
+		}
+
+		memcpy(base, snap_name, base_len);
+		base[base_len] = '\0';
 	}
 
-	memcpy(base, snap_name, base_len);
-	base[base_len] = '\0';
+	if (instance_key != NULL) {
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.c]: When `snap_name=""` this will call `memcpy(instance_key, NULL, 0);` I don't think this is a problem but then again this function doesn't promise to be called with only validated `snap_name`

File: cmd/libsnap-confine-private/snap-test.c [hunk 4]
```diff
@@ -287,6 +287,80 @@ static void test_sc_snap_drop_instance_name_basic(void)
...
+				    sizeof instance);
+	g_assert_cmpstr(name, ==, "foo");
+	g_assert_cmpstr(instance, ==, "");
+
+	memset(name, 0xff, sizeof name);
+	sc_snap_split_instance_name("foo_bar", name, sizeof name, NULL, 0);
+	g_assert_cmpstr(name, ==, "foo");
+
+	memset(instance, 0xff, sizeof instance);
+	sc_snap_split_instance_name("foo_bar", NULL, 0, instance,
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap-test.c]: Can you add two more tests please, with `snap_name` having value `""` and `"hello_world_surprise"` please.

Comment (bboozzoo) [reply cmd/libsnap-confine-private/snap-test.c]: added

File: cmd/libsnap-confine-private/snap.h (hunk 2 above)
Comment (bboozzoo) [reply cmd/libsnap-confine-private/snap.h]: renamed to: 
```
void sc_snap_split_instance_name(const char *name, char *snap_name,
			 size_t snap_name_size, char *instance_key, size_t instance_key_size);
```

File: cmd/libsnap-confine-private/snap.c [hunk 5]
```diff
@@ -169,25 +169,49 @@ void sc_snap_name_validate(const char *snap_name, struct sc_error **errorp)
 	sc_error_forward(errorp, err);
 }
 
-void sc_snap_drop_instance_name(const char *snap_name, char *base,
-				size_t base_size)
+void sc_snap_drop_instance_name(const char *name, char *snap_name,
```
Comment (niemeyer) [inline cmd/libsnap-confine-private/snap.c]: Below the function is using "instance name" as the composition of a snap name and an instance key, which seems nice. But in that case we cannot say "drop instance name" here, as we'd have nothing left. This should say "drop_instance_key" instead.

File: cmd/libsnap-confine-private/snap.c [hunk 6]
```diff
@@ -169,25 +169,49 @@ void sc_snap_name_validate(const char *snap_name, struct sc_error **errorp)
...
-		die("base or snap name unset");
+	sc_snap_split_instance_name(name, snap_name, snap_name_size, NULL, 0);
+}
+
+void sc_snap_split_instance_name(const char *name, char *snap_name,
+				 size_t snap_name_size, char *instance_key,
+				 size_t instance_key_size)
+{
+	if (name == NULL) {
+		die("name not provided");
```
Comment (niemeyer) [inline cmd/libsnap-confine-private/snap.c]: See note below.

File: cmd/libsnap-confine-private/snap.c [hunk 7]
```diff
@@ -169,25 +169,49 @@ void sc_snap_name_validate(const char *snap_name, struct sc_error **errorp)
...
+	}
+	if (snap_name == NULL && instance_key == NULL) {
+		die("snap name and instance key are unset");
```
Comment (niemeyer) [inline cmd/libsnap-confine-private/snap.c]: This should say something like:

"internal error: cannot split instance name when both snap name and instance key are unset"

and in the earlier case:

"internal error: cannot split snap name when snap name is unset"

Let's please keep an eye on having error messages that the user can relate to (and developers too, in fact).

File: cmd/libsnap-confine-private/snap.c [hunk 8]
```diff
@@ -169,25 +169,49 @@ void sc_snap_name_validate(const char *snap_name, struct sc_error **errorp)
...
+	size_t snap_name_len = 0;
+	size_t instance_key_len = 0;
 	if (pos == NULL) {
-		base_len = strlen(snap_name);
+		snap_name_len = strlen(name);
 	} else {
-		base_len = pos - snap_name;
+		snap_name_len = pos - name;
+		instance_key_start = pos + 1;
+		instance_key_len = strlen(name) - snap_name_len - 1;
```
Comment (niemeyer) [inline cmd/libsnap-confine-private/snap.c]: instance_key_len = strlen(instance_key_start);

?

File: cmd/libsnap-confine-private/snap.c [hunk 9]
```diff
@@ -169,25 +169,49 @@ void sc_snap_name_validate(const char *snap_name, struct sc_error **errorp)
...
+		die("name not provided");
+	}
+	if (snap_name == NULL && instance_key == NULL) {
+		die("snap name and instance key are unset");
 	}
 
-	const char *pos = strchr(snap_name, '_');
-	size_t base_len = 0;
+	const char *pos = strchr(name, '_');
+	const char *instance_key_start = NULL;
```
Comment (niemeyer) [inline cmd/libsnap-confine-private/snap.c]: This is a safer default (see note from zyga below):

const char *instance_key_start = "";

File: cmd/libsnap-confine-private/snap.c (hunk 3 above)
Comment (niemeyer) [reply cmd/libsnap-confine-private/snap.c]: We should definitely avoid that case, and it's easy to do so per note above.

File: cmd/libsnap-confine-private/snap.h [hunk 10]
```diff
@@ -60,15 +60,31 @@ bool verify_security_tag(const char *security_tag, const char *snap_name);
...
  *
  * For example: snap_instance => snap, just-snap => just-snap
  **/
-void sc_snap_drop_instance_name(const char *snap_name, char *base,
-				size_t base_size);
+void sc_snap_drop_instance_key(const char *name, char *snap_name,
+			       size_t snap_name_size);
+
+/**
+ * Extract snap name and instance key out of a snap with optional instance name string.
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.h]: > <Chipaca> zyga: AIUI instance name == name + "_" + key
> <mborzecki> zyga: foo_bar, foo_bar === instance name, foo === snap name, bar === instance key

With that said please reword this as follows:
```
Extract snap name and instance key out of an instance name.

A snap may be installed multiple times in parallel under distinct instance
names. This function extracts the snap name and instance key out of the instance
name. One of snap_name, instance_key must be non-NULL.

For example:
   name_instance => "name" & "instance"
   just-name     => "just-name" & ""
```

File: cmd/libsnap-confine-private/snap.h [hunk 11]
```diff
@@ -60,15 +60,31 @@ bool verify_security_tag(const char *security_tag, const char *snap_name);
 bool sc_is_hook_security_tag(const char *security_tag);
 
 /**
- * Extract snap name out of a snap with optional instance name string.
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.h]: This is fine

File: cmd/libsnap-confine-private/snap.h [hunk 12]
```diff
@@ -60,15 +60,31 @@ bool verify_security_tag(const char *security_tag, const char *snap_name);
...
  * A snap may be installed multiple times in parallel under distinct instance names.
  * This function extracts the snap name out of a name that possibly contains a snap
- * instance name.
+ * instance key.
  *
  * For example: snap_instance => snap, just-snap => just-snap
  **/
-void sc_snap_drop_instance_name(const char *snap_name, char *base,
-				size_t base_size);
+void sc_snap_drop_instance_key(const char *name, char *snap_name,
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.h]: Please change the first argument to `const char *instance_name`

File: cmd/libsnap-confine-private/snap.h [hunk 13]
```diff
@@ -60,15 +60,31 @@ bool verify_security_tag(const char *security_tag, const char *snap_name);
...
+ * A snap may be installed multiple times in parallel under distinct instance
+ * names. This function extracts the snap name and instance key out of a name
+ * that possibly contains a snap instance name. One of snap_name,
+ * instance_key must be non-NULL.
+ *
+ * For example:
+ *    snap_instance => "snap"      & "instance"
+ *    just-snap     => "just-snap" & ""
+ **/
+void sc_snap_split_instance_name(const char *name, char *snap_name,
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.h]: Please rename the first argument to `const char *instance_name`

File: cmd/libsnap-confine-private/snap.c [hunk 14]
```diff
@@ -169,25 +169,49 @@ void sc_snap_name_validate(const char *snap_name, struct sc_error **errorp)
...
-		die("base or snap name unset");
+	sc_snap_split_instance_name(name, snap_name, snap_name_size, NULL, 0);
+}
+
+void sc_snap_split_instance_name(const char *name, char *snap_name,
+				 size_t snap_name_size, char *instance_key,
+				 size_t instance_key_size)
+{
+	if (name == NULL) {
+		die("internal error: cannot split snap name when snap name is unset");
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.c]: Perhaps accordingly `cannot split instance name when it is unset`
